tests/test_unmarked.py

```diff
@@ -75,41 +75,34 @@ def test_unmarked(self):
           node
         )
 
-    self.assertIsNone(atok._tokens)
+    self.assertIsNone(atok._asttokens)
 
     has_tokens = False
     for node in ast.walk(atok.tree):
       atok_text = atok.get_text(node, padded=True)
       ast_text = ast.get_source_segment(source, node, padded=True)
       if ast_text:
-        if (
-            ast_text.startswith("f") and isinstance(node, (ast.Str, ast.FormattedValue))
-            or is_fstring_format_spec(node)
-        ):
-          self.assertEqual(atok_text, "", node)
-        else:
-          self.assertEqual(atok_text, ast_text, node)
+        self.assertEqual(atok_text, ast_text, node)
```

Comment:
The issue is that `ast_text` is a mess in the first place. Here's a demo:

```python
import ast

source = 'f"a{b:{c:d}}"'

tree = ast.parse(source)
for node in ast.walk(tree):
  if ast.get_source_segment(source, node) == source:
    print(ast.dump(node))
```

Output:

```python
Expr(value=JoinedStr(values=[Constant(value='a'), FormattedValue(value=Name(id='b', ctx=Load()), conversion=-1, format_spec=JoinedStr(values=[FormattedValue(value=Name(id='c', ctx=Load()), conversion=-1, format_spec=JoinedStr(values=[Constant(value='d')]))]))]))
JoinedStr(values=[Constant(value='a'), FormattedValue(value=Name(id='b', ctx=Load()), conversion=-1, format_spec=JoinedStr(values=[FormattedValue(value=Name(id='c', ctx=Load()), conversion=-1, format_spec=JoinedStr(values=[Constant(value='d')]))]))])
Constant(value='a')
FormattedValue(value=Name(id='b', ctx=Load()), conversion=-1, format_spec=JoinedStr(values=[FormattedValue(value=Name(id='c', ctx=Load()), conversion=-1, format_spec=JoinedStr(values=[Constant(value='d')]))]))
JoinedStr(values=[FormattedValue(value=Name(id='c', ctx=Load()), conversion=-1, format_spec=JoinedStr(values=[Constant(value='d')]))])
FormattedValue(value=Name(id='c', ctx=Load()), conversion=-1, format_spec=JoinedStr(values=[Constant(value='d')]))
JoinedStr(values=[Constant(value='d')])
Constant(value='d')
```

There's 6 AST nodes that all say that their source text spans the full f-string, when IMO none of them should. Does that seem OK to you?

Reply:
Ah, yeah, that's not great.
This feels like it would be worth some separate tests, rather than just bundled into the integration tests.

Reply:
Eek, yes, `get_source_segment` seems wrong, not only for the nodes for which it returns the full f-string, but also for nodes for which it doesn't. E.g. the first `Name` node, I imagine, should produce `'b'`, but is instead producing `'"'` (at least on Python 3.9.5)

Reply:
Oh, that's bad, well spotted. So yes, we'd need to do something about that too. I should also report this to Python if it hasn't been already. Here's a script:

```python
import ast

source = '''
f"a b { c + d : hgg } e f { g * 3 } h"
'''

tree = ast.parse(source)
for node in ast.walk(tree):
    segment = ast.get_source_segment(source, node)
    if not segment or segment.strip() == source.strip():
      continue
    print(ast.dump(node))
    print(segment)
    print()
```

Output:

```python
BinOp(left=Name(id='c', ctx=Load()), op=Add(), right=Name(id='d', ctx=Load()))
a b {

BinOp(left=Name(id='g', ctx=Load()), op=Mult(), right=Constant(value=3))
g * 3

Name(id='c', ctx=Load())
a

Name(id='d', ctx=Load())
{

Name(id='g', ctx=Load())
g

Constant(value=3)
3
```

Seems to be fixed in 3.10:

```python
BinOp(left=Name(id='c', ctx=Load()), op=Add(), right=Name(id='d', ctx=Load()))
c + d

BinOp(left=Name(id='g', ctx=Load()), op=Mult(), right=Constant(value=3))
g * 3

Name(id='c', ctx=Load())
c

Name(id='d', ctx=Load())
d

Name(id='g', ctx=Load())
g

Constant(value=3)
3
```